File: clients/android/app/src/main/kotlin/com/lextures/android/core/lms/CourseChecklistTargetTable.kt
package com.lextures.android.core.lms

import android.content.Context
import java.io.File

/** Bundled CC.8 native target table (anchor → Android destination). */
object CourseChecklistTargetTable {
    @Volatile
    private var cached: Map<String, String>? = null

    fun shared(context: Context? = null): Map<String, String> {
        cached?.let { return it }
        val map = load(context)
        cached = map
        return map
    }

    fun load(context: Context? = null): Map<String, String> {
        if (context != null) {
            runCatching {
                context.assets.open("checklist-targets.json").bufferedReader().use { it.readText() }
            }.getOrNull()?.let { return CourseChecklistLogic.loadTargetTable(it) }
        }
        // Tests / fallback: walk up to monorepo packages path.
        var dir = File(System.getProperty("user.dir") ?: ".")
        repeat(10) {
            val candidates = listOf(
                File(dir, "clients/packages/checklist-targets.json"),
                File(dir, "packages/checklist-targets.json"),
                File(dir, "app/src/main/assets/checklist-targets.json"),
            )
            for (f in candidates) {
                if (f.isFile) {
                    return CourseChecklistLogic.loadTargetTable(f.readText())
                }
            }
            dir = dir.parentFile ?: return emptyMap()
        }
        return emptyMap()
    }

    fun resetCacheForTests() {
        cached = null
    }
}

File: clients/ios/Lextures/Core/LMS/CourseChecklistTargetTable.swift
import Foundation

/// Bundled CC.8 native target table (anchor → iOS destination).
enum CourseChecklistTargetTable {
    private static var cached: [String: String]?

    static var shared: [String: String] {
        if let cached { return cached }
        let map = load()
        cached = map
        return map
    }

    static func load(bundle: Bundle = .main) -> [String: String] {
        if let url = bundle.url(forResource: "checklist-targets", withExtension: "json"),
           let data = try? Data(contentsOf: url) {
            return CourseChecklistLogic.loadTargetTable(from: data)
        }
        // Fallback: packages path when running tests without app resources.
        let thisFile = URL(fileURLWithPath: #filePath)
        let candidates = [
            thisFile
                .deletingLastPathComponent()
                .deletingLastPathComponent()
                .deletingLastPathComponent()
                .deletingLastPathComponent()
                .appendingPathComponent("packages/checklist-targets.json"),
            URL(fileURLWithPath: FileManager.default.currentDirectoryPath)
                .appendingPathComponent("clients/packages/checklist-targets.json"),
        ]
        for url in candidates {
            if let data = try? Data(contentsOf: url) {
                return CourseChecklistLogic.loadTargetTable(from: data)
            }
        }
        return [:]
    }

    #if DEBUG
    static func resetCacheForTests() {
        cached = nil
    }
    #endif
}
